### DISCUSSION TITLE
LLVM / Generate Rom Executables

### DISCUSSION THREAD

**Discussion by bdragoncore** [project64/project64#2017] [Ideas]
In recent works of my emulator I have been able to achieve performance by recompiling MIPS into Common Intermediate Language (CIL) while using Release / Ahead-of-time (AOT) compilation of the emulator itself.  I could easily write the emitted CIL into a separate assembly DLL that can cache all of the recompiled MIPS routines, this would allow AOT to be able to optimize them more.

Next I eventually want to pick up on the LLVM library and do the same.  Project64 could benefit from the same here via the LLVM library by generating LLVM IR from MIPS assembly.  Of course there is more overhead but this will let the recompiler be cross-platform ready based on the support of LLVM's official backends and everything can be easily cached into a file.  For each rom provided, an executable could be generated, and maybe eventually shaders or SPIR-V (Vulkan) can be cached into it as well.  Anything that was not yet hit, will need a way to update the cache in the file.

Also I wonder about using the cloud to cache all possible LLVM IR blocks using various hash lookups, and that it could optimize automatically on the server per CPU architecture / deep learning, etc.  I think when each rom offline executable has been generated, a static block list is fetched and compiled AOT.  If self-modifying code is predicable, that could be handled as well.

**Comment by JunielKatarn** [comment]
@bryanperris Have you had any progress with this experiment?

**Reply by bdragoncore** [reply]
Not recently, I have been focused at trying to understand the RSP and Fast3D v1.
